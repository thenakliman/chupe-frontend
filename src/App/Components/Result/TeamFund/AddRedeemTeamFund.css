.team-fund-add-redeem-container {
  display: flex;
  flex-direction: column;
  position: fixed;
  margin: 3% 20%;
  border: 1px solid #000;
  width: 30rem;
  height: 25rem;
}

.add-redeem-team-fund-select {
  display: flex;
  justify-content: space-evenly;
  padding: 1.5rem 5rem;
}

.add-redeem-input-amount {
  display: flex;
  justify-content: space-evenly;
  padding: 1.5rem 5rem;
}

.team-fund-amount-container {
  display: flex;
  justify-content: space-evenly;
  padding: 1.5rem 5rem;

}

.team-fund-submit-button {
  display: flex;
  justify-content: space-evenly;
  padding: 1.5rem 5rem;
}

.team-fund-amount {
  width: 10rem;
}

.team-fund-button {
  padding: 1rem 2rem;
  margin: 2rem 0;
  float: right;
  color: white;

  cursor: pointer;
}