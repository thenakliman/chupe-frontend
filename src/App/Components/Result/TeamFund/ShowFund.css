.team-fund-container {
  width: 65rem;
  margin: 0 auto;
}