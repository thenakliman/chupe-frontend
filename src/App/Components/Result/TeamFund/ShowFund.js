import PropTypes from 'prop-types';
import React from 'react';

/**
* Show Fund component of the application.
*
* @author: thenakliman
*/
export class ShowFund extends React.Component {
  /** Fetch team fund while mounting this component */
  componentWillMount() {
    this.props.fetchFunds(this.props.match.params.id);
  }

  /** Component for show funds.
   * @return {object} Return component for showing funds
   */
  render() {
    return (
      <div id='user-fund-container-id' className='team-fund-container'>
        <table id='fund-table-id'>
          <thead>
            <tr>
              <th>S.No</th>
              <th>Owner</th>
              <th>type</th>
              <th>Amount</th>
              <th>Added On</th>
            </tr>
          </thead>
          <tbody>
          { this.props.funds.map((fund, index) => (
              <tr key={`${fund.id}`}>
                <td>{index+1}</td>
                <td>{fund.owner}</td>
                <td>{fund.type}</td>
                <td>{'CREDIT' === fund.transactionType? fund.amount: -fund.amount}</td>
                <td>{fund.createdAt}</td>
              </tr>
            ))
          }
          </tbody>
        </table>
      </div>
    );
  }
}

ShowFund.propTypes = {
  fetchFunds: PropTypes.func.isRequired,
  funds: PropTypes.array,
};
