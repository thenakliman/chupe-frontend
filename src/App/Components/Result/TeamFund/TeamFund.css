.team-fund-container {
  margin: 3% 20%;
}