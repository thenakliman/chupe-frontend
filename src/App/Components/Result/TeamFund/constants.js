export const transactionMapping = {
  Add: 'CREDIT',
  Redeem: 'DEBIT',
};
