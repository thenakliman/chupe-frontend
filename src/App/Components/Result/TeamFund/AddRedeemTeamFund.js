import React from 'react';
import propTypes from 'prop-types';

require('./AddRedeemTeamFund.css');

/** Add or Redeem a team member fund. */
export class AddRedeemTeamFund extends React.Component {
  /** Constructor for the component.
   * @param {object} props for the component
   */
  constructor(props) {
    super(props);
    this.state = {
      fundType: '',
      amount: 0,
    };

    this.handleAmountUpdate = this.handleAmountUpdate.bind(this);
    this.handleFundType = this.handleFundType.bind(this);
    this.handleClick = this.handleClick.bind(this);
  }

  /** Fetch team fund type on mount. */
  componentWillMount() {
    this.props.fetchFundTypes();
  }

  /** Handle amount update.
   * @param {number} amount to be added
   */
  handleAmountUpdate(amount) {
    this.setState({amount: amount});
  }

  /** Handle fund type update.
   * @param {string} typeId
   */
  handleFundType(typeId) {
    const fund = this.props.fundTypes.find((fund) => fund.id == typeId);
    const defaultAmount = fund && fund.defaultAmount || 0;
    this.setState({fundType: typeId, amount: defaultAmount});
  }
  /** Handle submit event of button. */
  handleClick() {
    if (this.state.amount === 0) {
      return;
    }

    this.props.addFund(this.state.fundType, this.state.amount);
  }

  /** Return Add/Redeem team fund type component.
   * @return {object} component for add/redeem
   */
  render() {
    return (
      <div className={'team-fund-popup-container'}>
      <div id='team-fund-add-redeem-container-id'
           className='team-fund-add-redeem-container'>
        <div className='add-redeem-team-fund-select'>
          <select id='team-fund-select-type-id'
                  onChange={(event) => this.handleFundType(event.target.value)}
                  value={this.state.value}
                  >
            <option value=''>Select Fund Type</option>
            {
              this.props.fundTypes.map((fund) => (
                <option value={fund.id}
                        key={fund.id}>
                            {fund.description}
                </option>))
            }
          </select>
        </div>
        <div className='team-fund-amount-container'>
          Amount:
          <input id='team-fund-amount-id'
                 className='team-fund-amount'
                 value={this.state.amount}
                 onChange={
                  (event) => this.handleAmountUpdate(event.target.value)}
                 />
        </div>
        <div className='team-fund-submit-button'>
          <button id='team-fund-cancel-button-id'
                  className='team-fund-button'
                  onClick={(event) => this.props.closePopup()}>
                Cancel
          </button>
          <button id='team-fund-submit-button-id'
                  className='team-fund-button'
                  onClick={() => this.handleClick()}>
                Submit
          </button>
        </div>
      </div>
    </div>
    );
  }
}

AddRedeemTeamFund.propTypes = {
  fetchFundTypes: propTypes.func.isRequired,
  fundTypes: propTypes.arrayOf(propTypes.object).isRequired,
  addFund: propTypes.func.isRequired,
  closePopup: propTypes.func.isRequired,
};
