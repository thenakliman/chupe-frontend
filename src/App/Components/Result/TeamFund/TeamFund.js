import propTypes from 'prop-types';
import React from 'react';
import {history} from '../../../utils/history';
/* eslint-disable */
import {AddRedeemTeamFund} from './AddRedeemTeamFund';
/* eslint-enable */

require('./TeamFund.css');
/**
* Team Fund component of the application.
*
* @author: thenakliman
*/
export class TeamFund extends React.Component {
  /** Constructor, sets initial state and bind methods
   * @param {object} props initial state
   */
  constructor(props) {
    super(props);
    this.state = {hasPopup: false, owner: null, transactionType: ''};
    this.addRedeemFund = this.addRedeemFund.bind(this);
    this.onClickAdd = this.onClickAdd.bind(this);
    this.onClickRedeem = this.onClickRedeem.bind(this);
    this.addFund = this.addFund.bind(this);
  }
  /** Fetch team fund while mounting this component */
  componentWillMount() {
    this.props.fetchTeamFund();
  }

  /** toggle state of pop.
   * @param {string} owner
   * @param {string} transactionType
   */
  addRedeemFund(owner, transactionType) {
    let currentOwner;
    if (this.state.hasPopup) {
      currentOwner = null;
      transactionType = '';
    } else {
      currentOwner = owner;
    }

    this.setState({
        hasPopup: !this.state.hasPopup,
        owner: currentOwner,
        transactionType: transactionType,
    });
  }

  /** toggle state of pop.
   * @param {string} owner
   */
  onClickAdd(owner) {
    this.addRedeemFund(owner, 'CREDIT');
  }

  /** toggle state of pop.
   * @param {string} owner
   */
  onClickRedeem(owner) {
    this.addRedeemFund(owner, 'DEBIT');
  }

  /** toggle state of pop.
   * @param {string} owner
   */
  onClickDetail(owner) {
    history.push('/funds/'+owner);
  }

  /** Add fund.
   * @param {string} fundType
   * @param {number} amount
   */
  addFund(fundType, amount) {
    this.props.addFund({
      transactionType: this.state.transactionType,
      type: fundType,
      amount: amount,
      owner: this.state.owner,
      addedBy: this.props.loggedInUser,
      isApproved: false,
    });

    this.addRedeemFund(null);
  }

  /**
  * Team Fund result component of the application.
  *
  * @return {Object} UserResult component.
  */
  render() {
    return (
        <div id='team-fund-container-id' className='team-fund-container'>
          { this.state.hasPopup &&
            <AddRedeemTeamFund fetchFundTypes={this.props.fetchFundTypes}
                               addFund={this.addFund}
                               closePopup={() => this.addRedeemFund(null)}
                               fundTypes={this.props.fundTypes}/>
          }
          <table id='team-fund-table-id'>
            <thead>
              <tr>
                <th>S.No</th>
                <th>Username</th>
                <th>Amount</th>
              </tr>
            </thead>
            <tbody>
            { this.props.teamFund.map((teamMember, index) => (
                <tr key={`${teamMember.owner}`}>
                  <td>{index+1}</td>
                  <td>{teamMember.owner}</td>
                  <td>
                    <span className='user-amount-container'>
                      <span
                        className='fund-operation-negative-sign'
                        id={`fund-operation-negative-sign-${teamMember.owner}`}
                        onClick={() => this.onClickRedeem(teamMember.owner)}>
                        -
                      </span>
                      <span className={'user-fund'}>
                      <u id={`user-amount-${teamMember.owner}`}
                         onClick={() => this.onClickDetail(teamMember.owner)}>
                        <i className={'user-fund-amount'}>
                          {teamMember.totalAmount}
                        </i>
                      </u>
                      </span>
                      <span
                        className='fund-operation-positive-sign'
                        id={`fund-operation-positive-sign-${teamMember.owner}`}
                        onClick={() => this.onClickAdd(teamMember.owner)}>
                        +
                      </span>
                    </span>
                  </td>
                </tr>
              ))
            }
            </tbody>
          </table>
        </div>
    );
  }
}

TeamFund.propTypes = {
  teamFund: propTypes.arrayOf(propTypes.object).isRequired,
  fundTypes: propTypes.arrayOf(propTypes.object).isRequired,
  loggedInUser: propTypes.string.isRequired,
  fetchTeamFund: propTypes.func.isRequired,
  fetchFundTypes: propTypes.func.isRequired,
  addFund: propTypes.func.isRequired,
};
