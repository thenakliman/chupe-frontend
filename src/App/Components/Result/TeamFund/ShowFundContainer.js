import {ShowFund} from './ShowFund';
import * as fundAction from '../../../Actions/teamFundActions';
import {connect} from 'react-redux';

export const mapStateToProps = (state) => ({
  funds: state.funds,
});

export const mapDispatchToProps = (dispatch) => ({
  fetchFunds: (owner) => dispatch(fundAction.fetchFundsForAUser(owner)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ShowFund);
