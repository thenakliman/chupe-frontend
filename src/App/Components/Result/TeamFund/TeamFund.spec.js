import React from 'react'; // eslint-disable-line no-unused-vars
import {TeamFund} from './TeamFund'; // eslint-disable-line no-unused-vars
import {shallow, mount} from 'enzyme';
import toJson from 'enzyme-to-json';

describe('Team fund component snapshot', () => {
    it('should match the snapshot', () => {
        const wrapper = shallow(<TeamFund teamFund={[]}
                                          fundTypes={[]}
                                          loggedInUser={'username2'}
                                          addFund={()=>{}}
                                          fetchFundTypes={jest.fn()}
                                          fetchTeamFund={jest.fn()}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});

describe('Team Fund component', () => {
    it('Should have container div for Team fund', () => {
        const wrapper = shallow(<TeamFund teamFund={[]}
                                          fundTypes={[]}
                                          loggedInUser={'username2'}
                                          addFund={()=>{}}
                                          fetchFundTypes={jest.fn()}
                                          fetchTeamFund={jest.fn()}/>);

        expect(wrapper.find('#team-fund-container-id').length).toEqual(1);
    });

    it('Should have team fund table', () => {
        const wrapper = shallow(<TeamFund teamFund={[]}
                                          addFund={()=>{}}
                                          fundTypes={[]}
                                          loggedInUser={'username2'}
                                          fetchFundTypes={jest.fn()}
                                          fetchTeamFund={jest.fn()}/>);

        expect(wrapper.find('#team-fund-table-id').length).toEqual(1);
    });

    it('Should have table header for username', () => {
        const wrapper = shallow(<TeamFund teamFund={[]}
                                          addFund={()=>{}}
                                          fundTypes={[]}
                                          loggedInUser={'username2'}
                                          fetchFundTypes={jest.fn()}
                                          fetchTeamFund={jest.fn()}/>);

        const hasUsername = wrapper.find('th').someWhere(
            (header) => header.props().children === 'Username');
        expect(hasUsername).toBe(true);
    });

    it('Should have table header for S.No', () => {
        const wrapper = shallow(<TeamFund teamFund={[]}
                                          addFund={()=>{}}
                                          fundTypes={[]}
                                          loggedInUser={'username2'}
                                          fetchFundTypes={jest.fn()}
                                          fetchTeamFund={jest.fn()}/>);

        const hasUsername = wrapper.find('th').someWhere(
            (header) => header.props().children === 'S.No');
        expect(hasUsername).toBe(true);
    });

    it('Should have table header for Amount', () => {
        const wrapper = shallow(<TeamFund teamFund={[]}
                                          addFund={()=>{}}
                                          fundTypes={[]}
                                          loggedInUser={'username2'}
                                          fetchFundTypes={jest.fn()}
                                          fetchTeamFund={jest.fn()}/>);

        const hasUsername = wrapper.find('th').someWhere(
            (header) => header.props().children === 'Amount');
        expect(hasUsername).toBe(true);
    });

    it('Should have table header for Amount', () => {
        const wrapper = shallow(<TeamFund teamFund={[]}
                                          addFund={()=>{}}
                                          fundTypes={[]}
                                          loggedInUser={'username2'}
                                          fetchFundTypes={jest.fn()}
                                          fetchTeamFund={jest.fn()}/>);

        const hasUsername = wrapper.find('th').someWhere(
            (header) => header.props().children === 'Action');
        expect(hasUsername).toBe(true);
    });

    it('Should fetch team fund on component mount', () => {
        const fetchTeamFund = jest.fn();
        shallow(<TeamFund teamFund={[]}
                          addFund={()=>{}}
                          fundTypes={[]}
                          loggedInUser={'username2'}
                          fetchFundTypes={jest.fn()}
                          fetchTeamFund={fetchTeamFund} />);

        expect(fetchTeamFund).toHaveBeenCalledWith();
    });

    it('Should have table row for each team member', () => {
        const teamFund = [{
          username: 'username1',
          amount: 100,
        }, {
          username: 'username2',
          amount: -100,
        }];

        const wrapper = shallow(<TeamFund teamFund={teamFund}
                                          fundTypes={[]}
                                          addFund={()=>{}}
                                          loggedInUser={'username2'}
                                          fetchFundTypes={jest.fn()}
                                          fetchTeamFund={jest.fn()} />);

        expect(wrapper.find('tr').length ).toBe(3);
    });

    it('should have owner in state', () => {
        const teamFund = [{
          owner: 'username1',
          amount: 100,
        }, {
          owner: 'username2',
          amount: -100,
        }];

        const wrapper = shallow(<TeamFund teamFund={teamFund}
                                          fundTypes={[]}
                                          loggedInUser={'username2'}
                                          addFund={()=>{}}
                                          fetchFundTypes={jest.fn()}
                                          fetchTeamFund={jest.fn()} />);

        expect(wrapper.state().owner).toBe(null);
    });

    it('should have set owner state as per click', () => {
        const teamFund = [{
          owner: 'username1',
          amount: 100,
        }, {
          owner: 'username2',
          amount: -100,
        }];

        const wrapper = shallow(<TeamFund teamFund={teamFund}
                                          fundTypes={[]}
                                          loggedInUser={'username2'}
                                          addFund={()=>{}}
                                          fetchFundTypes={jest.fn()}
                                          fetchTeamFund={jest.fn()} />);

        wrapper.find('#add-redeem-team-fund-username2').simulate('click');
        expect(wrapper.state().owner).toBe('username2');
    });

    it('should have set owner state to null when pop is on', () => {
        const teamFund = [{
          owner: 'username1',
          amount: 100,
        }, {
          owner: 'username2',
          amount: -100,
        }];

        const wrapper = shallow(<TeamFund teamFund={teamFund}
                                          loggedInUser={'username2'}
                                          fundTypes={[]}
                                          addFund={()=>{}}
                                          fetchFundTypes={jest.fn()}
                                          fetchTeamFund={jest.fn()} />);
        wrapper.setState({hasPopup: true});
        wrapper.find('#add-redeem-team-fund-username2').simulate('click');
        expect(wrapper.state().owner).toBe(null);
    });

    it('should toggle hasPopUp to true from false', () => {
        const teamFund = [{
          owner: 'username1',
          amount: 100,
        }, {
          owner: 'username2',
          amount: -100,
        }];

        const wrapper = shallow(<TeamFund teamFund={teamFund}
                                          loggedInUser={'username2'}
                                          fundTypes={[]}
                                          addFund={()=>{}}
                                          fetchFundTypes={jest.fn()}
                                          fetchTeamFund={jest.fn()} />);

        wrapper.find('#add-redeem-team-fund-username2').simulate('click');
        expect(wrapper.state().hasPopup).toBe(true);
    });

    it('should toggle hasPopUp to false from true', () => {
        const teamFund = [{
          owner: 'username1',
          amount: 100,
        }, {
          owner: 'username2',
          amount: -100,
        }];

        const wrapper = shallow(<TeamFund teamFund={teamFund}
                                          loggedInUser={'username2'}
                                          addFund={()=>{}}
                                          fundTypes={[]}
                                          fetchFundTypes={jest.fn()}
                                          fetchTeamFund={jest.fn()} />);

        wrapper.find('#add-redeem-team-fund-username2').simulate('click');
        wrapper.find('#add-redeem-team-fund-username2').simulate('click');
        expect(wrapper.state().hasPopup).toBe(false);
    });

    it('should call addFund method', () => {
        const teamFund = [{
          owner: 'username1',
          amount: 100,
        }, {
          owner: 'username2',
          amount: -100,
        }];

        const addFund = jest.fn();
        const wrapper = mount(
            <TeamFund teamFund={teamFund}
                      addFund={addFund}
                      loggedInUser={'username2'}
                      fundTypes={[{type: 'BIRTHDAY', id: 1}]}
                      fetchFundTypes={jest.fn()}
                      fetchTeamFund={jest.fn()} />);

        wrapper.setState({hasPopup: true, owner: 'username1'});
        const amount = 1000;

        wrapper.find('#team-fund-select-type-id').simulate(
            'change', {target: {value: 1}});

        wrapper.find('#team-fund-action-select-id').simulate(
            'change', {target: {value: 'CREDIT'}});

        wrapper.find('#team-fund-amount-id').simulate(
            'change', {target: {value: amount}});
        wrapper.find('#team-fund-submit-button-id').simulate('click');

        expect(wrapper.state()).toEqual({hasPopup: false, owner: null});

        expect(addFund).toHaveBeenCalledWith(
          {
            addedBy: 'username2',
            amount: 1000,
            type: 1,
            isApproved: false,
            owner: 'username1',
            transactionType: 'CREDIT',
          }
        );
    });

    it('should call addFund method with default value of fundType', () => {
        const teamFund = [{
          owner: 'username1',
          amount: 100,
        }, {
          owner: 'username2',
          amount: -100,
        }];

        const fundTypes = [{type: 'BIRTHDAY', id: 1, defaultAmount: 101}];
        const addFund = jest.fn();
        const wrapper = mount(
            <TeamFund teamFund={teamFund}
                      addFund={addFund}
                      loggedInUser={'username2'}
                      fundTypes={fundTypes}
                      fetchFundTypes={jest.fn()}
                      fetchTeamFund={jest.fn()} />);

        wrapper.setState({hasPopup: true, owner: 'username1'});

        wrapper.find('#team-fund-select-type-id').simulate(
            'change', {target: {value: 1}});

        wrapper.find('#team-fund-action-select-id').simulate(
            'change', {target: {value: 'CREDIT'}});

        wrapper.find('#team-fund-submit-button-id').simulate('click');

        expect(wrapper.state()).toEqual({hasPopup: false, owner: null});

        expect(addFund).toHaveBeenCalledWith(
          {
            addedBy: 'username2',
            amount: 101,
            type: 1,
            isApproved: false,
            owner: 'username1',
            transactionType: 'CREDIT',
          }
        );
    });

    it('should change state on click of cancel button', () => {
        const teamFund = [{
          owner: 'username1',
          amount: 100,
        }, {
          owner: 'username2',
          amount: -100,
        }];

        const addFund = jest.fn();
        const wrapper = mount(
            <TeamFund teamFund={teamFund}
                      addFund={addFund}
                      loggedInUser={'username2'}
                      fundTypes={[{type: 'BIRTHDAY', id: 1}]}
                      fetchFundTypes={jest.fn()}
                      fetchTeamFund={jest.fn()} />);

        wrapper.setState({hasPopup: true, owner: 'username1'});
        wrapper.find('#team-fund-cancel-button-id').simulate('click');

        expect(wrapper.state()).toEqual({hasPopup: false, owner: null});
    });
});
