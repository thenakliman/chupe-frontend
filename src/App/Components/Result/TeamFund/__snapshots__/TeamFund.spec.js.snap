// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Team fund component snapshot should match the snapshot 1`] = `
<div
  id="team-fund-container-id"
>
  <table
    id="team-fund-table-id"
  >
    <thead>
      <tr>
        <th>
          S.No
        </th>
        <th>
          Username
        </th>
        <th>
          Amount
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
